### PR TITLE
Explore: Add error handling function when retrieving datasource

### LINKED ISSUE #54749: Explore: crash when last used datasource is deleted
When navigating to Explore, if the last used datasource was deleted, Grafana shows a blank page.
https://github.com/grafana/grafana/pull/51605 introduced this.

given most of the logic about initialization was moved to `ExplorePaneContainer`, in https://github.com/grafana/grafana/blob/38c1f3d054ef7ea1a86b41437004fdbe72505e3e/public/app/features/explore/ExplorePaneContainer.tsx#L83 we should fallback to the default datasource in case `initialDatasource`., which is roughly the same logic found in https://github.com/grafana/grafana/blob/0e13ee345d28b17a3a93696b2edf0685a29b4028/public/app/features/explore/state/utils.ts#L93, but it doesn't seem to be triggered.

### PR DESCRIPTION
**Pull request by gelicia**
Use function that gets and returns the datasource, with the error handling in case the datasource is not found. Also add a test for this.

Fixes #54749

### CODE REVIEW

**Comment by Elfo404** [comment]
there is still another case not working properly: in this case i had mixed enabled a 2 queries in explore (1 Loki and 1 Cloudwatch). i removed the Loki datasource and refreshed the page, this happened:

Screenshot 2022-09-13 at 10 40 24

Not sure wether we should fallback to the default DS or just remove the query, but most likely we shouldn't show "mixed" in the first query's DS picker.

**Comment by gelicia** [comment]
> Not sure wether we should fallback to the default DS or just remove the query, but most likely we shouldn't show "mixed" in the first query's DS picker.

Ahh yeah, this is caused by logic that looks for the root datasource to populate if the query datasource doesn't exist.

On first glance, the thing that makes the most sense to me is to remove the query with the deleted datasource and show an error about what happened. However, with previous logic I have it where if you go into explore with a link where the datasource is mixed but there is only one query with a datasource, the root datasource will change to that datasource. That's straightforward enough in the simple case, but with this case, you'd go from Mixed, Query A: 'Deleted DS', Query B: 'Valid DS' to Valid DS, Query B. Two rules affecting the input is kind of a lot.

I guess what I'm thinking is we go ahead with removing it and letting that happen, and see if we get any feedback from users. 

I just think that if I had a query with a deleted datasource, and that query changed to the default datasource, I'd be more confused than if it was just removed.

**Review by ifrost — APPROVED** [review]
Nice work 🎉

File: public/app/core/utils/explore.ts [hunk 1]
```diff
@@ -316,18 +316,30 @@ export async function ensureQueries(
...
+        }
+      }
+
+      if (validDS) {
+        allQueries.push({
+          ...query,
+          refId,
+          key,
+        });
+      }
```
[the remark this answers is not in the document]
Comment (gelicia) [reply public/app/core/utils/explore.ts:344]: > We don't need `validDS` guard here:

So the issue here is with compact URLs that don't have a query.datasource in them. In the old way, a query without a datasource would validate as true and keep going as usual, but with the update, a query without a datasource does not get pushed. 

I added a comment to explain why. Let me know if this logic still should be changed.

**Pull request merged**